### PR TITLE
Improve demo setup docs

### PR DESCRIPTION
**Pull request by bedengler**
The atk4/ui demo now runs also if you position it outside the atk4 repo and rename the "demos" directory to something else. This should add a layer of positive user experience, if someone tries that out and it works flawlessly.

In addition I have added documentation for & link to the atk4/ui demos.

### CODE REVIEW

File: docs/overview.rst [hunk 1]
```diff
@@ -451,3 +451,51 @@ infrastructure, Linux and SSH. A good place to start is Heroku (https://www.hero
 We reference Heroku in our tutorials, but Agile Toolkit can work with any cloud
 hosting that runs PHP apps.
 
+
+Online-Demo showing many functions
```
Comment (mvorisek) [inline docs/overview.rst]: this is ok if not written somewhere else alredy

File: docs/overview.rst [hunk 2]
```diff
@@ -451,3 +451,51 @@ infrastructure, Linux and SSH. A good place to start is Heroku (https://www.hero
...
+There's a demo available of atk4/ui which shows many of the modules and functions available
+in atk4.
+You can watch & use the source code of each example to find best practice examples and to see
+how to use atk4 in certain application cases.
+
+You can find the demo here:
+https://agile-ui.readthedocs.io/en/latest/index.html
+
+
+Demo contained in the atk4 repository
```
Comment (mvorisek) [inline docs/overview.rst]: text below is quite specifically worded...

there should be:
- advise demos need clone/fork
- how sqlite database can be created: `php demos/_demo-data/create-sqlite-db.php`

Comment (bedengler) [reply docs/overview.rst]: According to @georgehristov they are adviced to don't move the original repos and use them as they are.
What do we advice for consistency?  

1. clone just the "demos" directory or
2. clone the entire atk4/ui directory or
3. don't clone and use the files in /vendors/atk4/ui...

???

Comment (bedengler) [reply docs/overview.rst]: @mvorisek, I added SQLITE but didn't commit. I'm waiting for feedback for the cloning issue...

Comment (mvorisek) [reply docs/overview.rst]: fork whole atk4/ui + composer install + init sqlite db, then make whole atk4/ui accessible by webserver

Comment (bedengler) [reply docs/overview.rst]: What about moving just the "demos" directory in a subdirectory called "atk4/ui" on the webserver that's publicly available instead of the whole atk4/ui directory? 
Advantage: when updating atk4/ui, you have the demos separated and they aren't overwritten.

Comment (mvorisek) [reply docs/overview.rst]: impossible - you need to fork the whole ui

for separate vendor files, you need to understand composer more deeply and I think this should be not documented as too advanced for most users...

File: docs/overview.rst (hunk 1 above)
Comment (bedengler) [reply docs/overview.rst]: I didn't find it anywhere. Even if it's anywhere else, it should be placed in the overview (which is the first place when  someone is new to atk4 and opens the documentation).
And it should be removed from "anywhere else", if someone could find it.

File: docs/overview.rst (hunk 2 above)
Comment (bedengler) [reply docs/overview.rst]: adapted, please check

File: demos/init-app.php [hunk 3]
```diff
@@ -51,8 +51,13 @@
     throw new \atk4\ui\Exception('Database error: ' . $e->getMessage());
 }
 
-[$rootUrl, $relUrl] = preg_split('~(?<=/)(?=demos(/|\?|$))|\?~s', $_SERVER['REQUEST_URI'], 3);
-$demosUrl = $rootUrl . 'demos/';
+// Current sub directory of webserver - makes it independent where to run the demo from
+// basically we want to run it just from within the ui repo, but this adds a layer of
+// better user experience, when a user tries to copy it elsewhere and it works
+$currDir = substr(str_replace($_SERVER['DOCUMENT_ROOT'], "", dirname(__FILE__)), 1);
```
Comment (mvorisek) [inline demos/init-app.php]: I belive should be not supported, we should actually not distribute tests/demos in release at all for security reasons.

File: docs/overview.rst [hunk 4]
```diff
@@ -451,3 +451,71 @@ infrastructure, Linux and SSH. A good place to start is Heroku (https://www.hero
...
+ - Setup the database using the file "dump.sql" from the subdirectory "atk4/ui/demos/_demo-data" of the "demos" directory. You can also setup an SQLITE database using a provided script (see below)
+ - Open the demos from your browser (e.g. with https://localhost/vendors/atk4/ui/demos or if you choose to clone it, with e.g. https://localhost/atk4/ui/demos)
+
+Setup SQLITE database
+---------------------
+The demo also includes a script that let's you setup a SQLITE database. You will find this script in the subdirectory
+"atk4/ui/demos/_demo-data/".
+Call it as follows (you may have to adapt the paths to fit your webserver structure)::
+
+    php atk4/ui/demos/_demo-data/create-sqlite-db.php;  // when using the CLI approach
```
Comment (mvorisek) [inline docs/overview.rst]: we should allow to create from CLI only

File: docs/overview.rst [hunk 5]
```diff
@@ -451,3 +451,71 @@ infrastructure, Linux and SSH. A good place to start is Heroku (https://www.hero
...
+
+or::
+
+    https://localhost/atk4/ui/demos/_demo-data/create-sqlite-db.php;  // when calling in the browser
+
+
+Additional tips
+---------------
+
+ - If you are using Eclipse, import the previously cloned "atk4" directory into Eclipse (then it finds all sub-projects)
```
Comment (mvorisek) [inline docs/overview.rst]: no specific doc for Eclipse

File: demos/init-app.php (hunk 3 above)
Comment (bedengler) [reply demos/init-app.php]: I'm seeing it from the perspective as a user (I for myself joined atk4 a year ago). 
The learning curve is ridiculous steep as there's a big lack of documentation. Reading everything out of the code without having examples makes everything even more time consuming. Which is unnecessary imo, if we just 
1) offer more examples
2) write documentations
3) make it as easy as possible to get results (raise usability)

Currently you have to try out a lot. 
If you just spend hours or days of getting a demo running, users jump off again before they even were able to see a welcome screen...

Therefore I'd be interested in opinions of other team members first

Comment (mvorisek) [reply demos/init-app.php]: please post usecase (step by step) that needs this change

**Comment by mvorisek** [comment]
@bedengler revert `demos/init-app.php`, then we @ibelar or @georgehristov can review or we can close completely

File: demos/init-app.php (hunk 3 above)
Comment (romaninsh) [reply demos/init-app.php]: I think this is a positive change and it does not break anything. Sound reasoning on moving things outside.

File: docs/overview.rst [hunk 6]
```diff
@@ -451,3 +451,71 @@ infrastructure, Linux and SSH. A good place to start is Heroku (https://www.hero
...
+https://agile-ui.readthedocs.io/en/latest/index.html
+
+
+Demo contained in the atk4 repository
+=====================================
+When you download and install atk4 you will find a subdirectory called "demos" in the atk4
+repository which also could be locally executed.
+This folder is basically intended to run under the UI package only.
+That means the demo may not work properly, if you copy the "demos" directory out of this location
+(to run it from a different location) or if the "vendors" directory isn't accessible through the
```
Comment (romaninsh) [inline docs/overview.rst]: I thought the above fix addresses that? 

Could you rephrase this text in the form of step-by-step instruction on how to actually run demo elsewhere?

File: docs/overview.rst (hunk 4 above)
Comment (romaninsh) [reply docs/overview.rst]: it might be more confusing to offer alternative implementation, but if your opinion as someone new to ATK that both instruction should be provided -that's OK.. 

Just consider file permission implications. 12-factor app requires applications to be runnable in read-only containers.

Same can be solved with docker-compose and mysql container, btw.

File: docs/overview.rst (hunk 5 above)
Comment (romaninsh) [reply docs/overview.rst]: in my case directory is always called "ui", not atk4.

**Comment by romaninsh** [comment]
@bedengler - thanks so much for not only identifying documentation issue, but also be willing to improve the situation. Framework relies on contributors and we need more people focused specifically on the docs, so if you are serious about this - you would be able to refine documentation gradually.

Overall - your instruction seem very "specific" to your personal use case, can require a bit more work, but having this much is better than having nothing at all, so well done!

File: demos/init-app.php (hunk 3 above)
Comment (mvorisek) [reply demos/init-app.php]: > 
> 
> please post usecase (step by step) that needs this change

as written, I need usecase that needs this change, otherwise `str_replace($_SERVER['DOCUMENT_ROOT'], "", dirname(__FILE__)` seems super hacky to me and it will not work on even slighly badly configured servers!

**Review by mvorisek — CHANGES_REQUESTED** [review]
need changes

**Comment by mvorisek** [comment]
@bedengler Please proof read the changes, remove what is not needed/up-to-date and we can merge then.

**Review by abbadon1334 — APPROVED** [review]
LGTM

**Review by mvorisek — APPROVED** [review]
